### DISCUSSION TITLE
Permit applying user-defined conversions to constant patterns

### DISCUSSION THREAD

**Discussion by RikkiGibson** [dotnet/csharplang#6597] [Language Ideas]
Related to https://github.com/dotnet/roslyn/issues/63476

The motivating scenario is the following: [SharpLab](https://sharplab.io/#v2:EYLgtghglgdgPgAQAwAIEEYB0ANMAbTAGVgEcBuAWAChrsA5CMAUxRkZYF4V73MBxJgBcAFACIIogJSUaVKADMUwtsxQcu40SjhxW7NRuBSUAehMoA9gGtqAb2oBfatQVKVLKAGcUmywCcfI0lTcyY/Pws/EBQAQRQAYwsYT0EIGEEUADcIPABXD28mAA8AByZ4wSYAEztHZ1kgA)
```cs
using System.Xml.Linq;

XName name = XName.Get("a");

if (name == "a" || name == "b") // ok
{
}

if (name is "a" or "b") // error: A constant value is expected
{
}
```

`XName` declares an implicit conversion to `string`, which allows users to compare them directly to strings. But we don't allow this in patterns, because the result of a user-defined conversion is not a constant value. This could mean that the conversion has a nondeterministic result, and the optimizations in patterns which avoid performing the same tests multiple times are invalid. For example, `(name, age) is ("a", 42) or ("a", 43)` might need to actually test against `(XName)"a"` each time it occurs lexically.

This starts to feel like an unnecessary defense against a pathological scenario. We *do not* defend against things like this with properties--in a scenario like `(a,b) is ({Prop:true},1) or ({Prop:true},2)` we only test the property once in practice, and we "reserve the right" to test the property any number of times, assuming that within the pattern it will give a deterministic result.

It seems like conversion operators should be assumed to be deterministic in the same way and therefore allowed to be used with constant patterns. It would be useful to see any examples in real-world types where they are not deterministic. We would also treat such patterns as "handling" the constant values in the natural type--so that we can continue to "know" that `name is "a" or "a"` is redundant, for example. Effectively we would assume a 1:1 mapping between the inputs and outputs of a user-defined conversion.

**Comment by HaloFour** [comment]
Sounds like something better tackled with active patterns?

**Reply by RikkiGibson** [reply]
It seems difficult to compare which would be better without seeing more specific usage examples, the onboarding story for type owners (who can declare an active pattern with a certain type of input?), etc.

**Comment by 333fred** [comment]
One of our main concerns in preventing this in the first place was that this further exacerbates the issue of the statically-known input type changing the result of a pattern match. For example:

```cs
name is "a" // True
((object)name) is "a" // False, no compiler error
```

**Reply by qrli** [reply]
Isn't it also true for
```cs
if (((object)name) == "a" || ((object)name) == "b") // False, no compiler error
```

**Reply by 333fred** [reply]
It is true, but let's consider a similar example with regular strings:

```cs
var s1 = "sss";
object s2 = new string('s', 3);
Console.WriteLine(s1 == s2); // False, and also a warning that you're not doing string comparison
Console.WriteLine(s2 is "sss"); // True
```

When you explicitly cast to `object` and use `==`, you're intentionally using `ReferenceEquals`. Whereas pattern matching intentionally does type-checking. That difference is what causes our hesitation here: whether an instance of a `string` is statically typed as `string` or `object`, `is "str"` will work.

**Reply by RikkiGibson** [reply]
Another scenario where the result depends on static type of the input is `(object)1L is 1`.

**Reply by 333fred** [reply]
Yep. And we explicitly blocked using numeric constant patterns on type parameters constrained to `INumberBase<T>` to avoid spreading that further, and to make it so that a future language version can make it "just work™"

**Reply by RikkiGibson** [reply]
Is there a general idea of how the right solution to that would work? Like if a pattern input is not a sealed type, you might call some virtual method or do some type test to decide what method of comparison to use? It seems like that test could be user-defined which basically addresses the original scenario.

**Reply by 333fred** [reply]
My thinking had been specific to INumberBase<T>, not generalized.

**Reply by qrli** [reply]
I think it could be made a little more explicit:
```cs
if (((object)name) is XName "a" or XName "b")
```
Then static typing of left operand is no longer relevant, and the right operand still has a static type like the old `is`, while taking user defined conversions.

**Comment by qrli** [comment]
My experience is that this is more common with `JValue`, `JsonNode`, `BsonValue`, etc.

Not sure if I really want `is` to take user defined conversions, as `is` has traditionally meant to test for the type on the right side. If it takes user defined conversions, the type on the right will not be trustable any more. I'd vote against it in this case.

However, for `switch` case, I'd feel like it to support user defined conversions, which would mean inconsistent behavior than `is`. I'm torn.
